src/AutoMapper/Internal/FeatureCollectionBase.cs

```diff
@@ -0,0 +1,73 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+
+namespace AutoMapper.Internal
+{
+    public class FeatureCollectionBase<TValue> : IEnumerable<KeyValuePair<Type, TValue>>
+    {
+        private IDictionary<Type, TValue> _features = new Dictionary<Type, TValue>();
+
+        public TValue this[Type key]
+        {
+            get
+            {
+                return _features.TryGetValue(key, out var feature)
+                    ? feature
+                    : default(TValue);
+            }
+            set
+            {
+                if (value == null)
+                {
+                    _features.Remove(key);
+                }
+                else
+                {
+                    _features[key] = value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets the feature of type <typeparamref name="TFeature"/>.
+        /// </summary>
+        /// <typeparam name="TFeature">The type of the feature.</typeparam>
+        /// <returns>The feature or null if feature not exists.</returns>
+        public TFeature Get<TFeature>()
+            where TFeature : TValue
+        {
+            return (TFeature)this[typeof(TFeature)];
+        }
+
+        /// <summary>
+        /// Sets the feature for type <typeparamref name="TFeature"/>.
+        /// </summary>
+        /// <typeparam name="TFeature">The type of the t feature.</typeparam>
+        /// <param name="feature">The feature. To remove a feature, pass null value.</param>
+        public void Set<TFeature>(TFeature feature)
```

Comment:
This doesn't need to be generic. We can rely on GetType and this would simplify things, as the key/value thing is encapsulated and all calling code can just use the values.

Reply:
Is there any case we'd prefer `TFeature` over `feature.GetType()`? I can't think of any.

Reply:
I think `typeof(TFeature)` will be treated more as compile time instead of running `feature.GetType()` but I can be wrong on that. The usage will be the same with `Fetures.Set(new MyFeatureClass())`.

Reply:
The `typeof(TFeature)` is using the declared type and the `feature.GetType()` is using the runtime type (https://stackoverflow.com/a/17495443). So if we using `TFeature` we can allow sub-classing of the feature-classes and they still will be treated as the same. Using `.GetType()` they will be different individual features.

From that I prefer the typed version.

Reply:
Yes, but we don't have to support that here. Extensions developers inherit away internally.

Reply:
If we see the typed version with `TFeature` we can restrict the feature to be of a specific type. If we use `feature.GetType()` we also need an `if (!(feature is TValue)) throw new ArgumentException("Feature is of wrong type.");` to ensure the type of the feature. I prefer to have the typed version.